src/main/java/be/syntra/devshop/DevshopBack/security/jwt/JWTTokenProvider.java

```diff
@@ -0,0 +1,94 @@
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.stereotype.Component;
+
+import java.security.Key;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Date;
+import java.util.stream.Collectors;
+
+@Component
+@Slf4j
+public class JWTTokenProvider implements InitializingBean {
+
+    private static final String AUTHORITIES_KEY = "auth";
+    private static final String USER_ID = "userId";
+    private final String base64Secret;
+    private final Long tokenValidityInMilliseconds;
+    public Key key;
+
+    public JWTTokenProvider(
+            @Value("${jwt.base64-secret}") String base64Secret,
+            @Value("${jwt.token-validity-in-seconds}") Long tokenValidityInSeconds) {
+        this.base64Secret = base64Secret;
+        this.tokenValidityInMilliseconds = tokenValidityInSeconds * 1200;
+    }
+
+    @Override
+    public void afterPropertiesSet() {
+        byte[] keyBytes = Decoders.BASE64.decode(base64Secret);
+        this.key = Keys.hmacShaKeyFor(keyBytes);
+    }
+
+    public String createToken(Authentication authentication) {
+        Long now = System.currentTimeMillis();
+        Date validity = new Date(now + this.tokenValidityInMilliseconds);
+        return Jwts.builder()
+                .setSubject(authentication.getName())
+                .claim(AUTHORITIES_KEY, authentication.getAuthorities())
+                .setHeaderParam(USER_ID, ((CustomUser) authentication.getPrincipal()).getUserID())
+                .signWith(key, SignatureAlgorithm.HS512)
+                .setExpiration(validity)
+                .compact();
+    }
+
+    public Authentication getAuthentication(String token) {
+        Claims claims = Jwts.parser()
+                .setSigningKey(key)
+                .parseClaimsJws(token)
+                .getBody();
+        Collection<? extends GrantedAuthority> authorities =
+                Arrays.stream(claims.get(AUTHORITIES_KEY).toString().split(","))
+                        .map(SimpleGrantedAuthority::new)
+                        .collect(Collectors.toUnmodifiableList());
+        Long gebruikerId = null != claims.get(USER_ID)
+                ? (Long) claims.get(USER_ID)
+                : null;
+        CustomUser principal = new CustomUser(claims.getSubject(), "", authorities, gebruikerId);
+
+        return new UsernamePasswordAuthenticationToken(principal, token, authorities);
+    }
+
+    public boolean validateToken(String authToken) {
+        try {
+            Jwts.parser().setSigningKey(key).parseClaimsJws(authToken);
+            return true;
+        } catch (io.jsonwebtoken.security.SecurityException | MalformedJwtException e) {
```

Comment:
can this be extracted into a method?

Reply:
this is an excisting securityException. The reason I had to enclose the full import in the code, is because Java doens't allow duplicate import paths. I rather not write a customException deriving from this one for fear of breaking the code. It's also bad practice to merge two Exceptions into one